audbcards/sphinx/__init__.py

```diff
@@ -0,0 +1,98 @@
+import sphinx
+
+import audb
+import audeer
+
+from audbcards.core.datacard import Datacard
+from audbcards.core.dataset import Dataset
+from audbcards.core.dataset import create_datasets_page
+
+
+__version__ = '0.1.0'
+
+
+# ===== MAIN FUNCTION SPHINX EXTENSION ====================================
+def setup(app: sphinx.application.Sphinx):
+    r"""Modelcard Sphinx extension."""
+    app.add_config_value('audbcards_output_path', 'datasets', False)
```

Comment:
I am seing a problem here when I build the documentation
using 

    python     -m sphinx docs/ build/sphinx/html -b html                                                                                                                                                                                                                                                         

```

This is the trace:

Running Sphinx v5.3.0                                                                                                                                                                                                                                                                                                                                                                       
loading intersphinx inventory from https://audeering.github.io/audb/objects.inv...                                                                                                                                                                                                                                                                                                          
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                      
Parse emodb-1.4.1... /data/work3/cgeng/work/tools-public/audbcards/audbcards/core/dataset.py:293: 
... wrote /data/work3/cgeng/work/tools-public/audbcards/docs/data-public/emodb.rst                                                                                                                                                                                                                                                                                                          
done                                                                                                                                                                                                                                                                                                                                                                                        
current_repos                                                                                                                                                                                                                                                                                                                                                                               
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                      
current_repos                                                                                                                                                                                                                                                                                                                                                                               
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                      
Parse emodb-1.4.1...                                                                                                                                                                                                                                                                                                                                                                        
Extension error (audbcards.sphinx):                                                                                                                                                                                                                                                                                                                                                         
Handler <function builder_inited at 0x7f9338f9a430> for event 'builder-inited' threw an exception (exception: An exception was raised by the backend, please see stack trace for further information.)      
```

When I set a breakpoint, I can run the  `app.connect` invocations:

```
(Pdb) app.connect('builder-inited', builder_inited)                                                                                                                                                                                                                                                                                                                                         
70                                                                                                                                                                                                                                                                                                                                                                                          
(Pdb) app.connect('build-finished', builder_finished)                                                                                                                                                                                                                                                                                                                                       
71                                                                                                                                                                                                                                                                                                                                                                                          
(Pdb) c                                                                                                                                                                                                                                                                                                                                                                                     
loading intersphinx inventory from https://audeering.github.io/audb/objects.inv...                                                         
```

I cannot figure out a way to get more useful output (simply doing `python -v` does also not help).

Reply:
That's in general a problem with the sphinx extensions.
You can try to run:
```bash
$ python -m sphinx docs/ build/html -b html -vvv
```

Reply:
Which Python version do you use?

Reply:
> An exception was raised by the backend, please see stack trace for further information.

In most cases this indicates that there are some connections or rights issues with the backend (here repo on Artifactory). But I'm not sure as I doubt you have access problems.

Reply:
I added a CI job that builds the documentation and it fails when parsing the empty repo with private access rights: https://github.com/audeering/audbcards/actions/runs/7169842181/job/19521158928

Maybe you experience the same error?

Reply:
I will need to figure out was is going on here and how to fix it tomorrow.

Reply:
I can reproduce it in the venvs on both machines that I am using, one has a Python `3.8.10`, the other `3.9.14`

Reply:
> Which Python version do you use?

I can reproduce it in the venvs on both machines that I am using, one has a Python `3.8.10`, the other `3.9.14`

Reply:
> I added a CI job that builds the documentation and it fails when parsing the empty repo with private access rights: https://github.com/audeering/audbcards/actions/runs/7169842181/job/19521158928
> 
> Maybe you experience the same error?

The exception messages differ a bit:

I am getting locally:
```
Handler <function builder_inited at 0x7f9338f9a430> for event 'builder-inited' threw an exception (exception: An exception was raised by the backend, please see stack trace for further information.)
```
This is what the CI job says:
```
Handler <function builder_inited at 0x7f80d1412050> for event 'builder-inited' threw an exception (exception: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')))
```

Reply:
> > An exception was raised by the backend, please see stack trace for further information.
> 
> In most cases this indicates that there are some connections or rights issues with the backend (here repo on Artifactory). But I'm not sure as I doubt you have access problems.

I never made any configurations for  https://audeering.jfrog.io/artifactory  - or is there a API key needed for this?

Reply:
No, the idea is that it works with anonymous access, so you don't need to configure Artifactory for it.

Reply:
Good to know. I have used the `-vvvv ` flag at the very end and this gives more information.

From `conf.py`, when I comment out  
- the `all-repos` tuple it breaks.
- the `data-private` tuple it is still working

And: 

it is related to a `401 unauthorized`  in the backend.

This is the 401 when raised:

```
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
Parse emodb-1.4.1... /data/work3/cgeng/work/tools-public/audbcards/audbcards/core/dataset.py:293: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.                                                                                                                                                                                                                                                                                                                      
  filter = data.applymap(lambda d: d == [])                                                                                                                                                                                                                                                                                                                                                                                                                                                              
... wrote /data/work3/cgeng/work/tools-public/audbcards/docs/data-public/emodb.rst                                                                                                                                                                                                                                                                                                                                                                                                                       
done                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
Get list of available datasets... done                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
Parse emodb-1.4.1...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
  File "/data/work3/cgeng/.envs/audbcards2/lib/python3.9/site-packages/dohq_artifactory/exception.py", line 20, in raise_for_status                                                                                                                                                                                                                                                                                                                                                                      
    response.raise_for_status()                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
  File "/data/work3/cgeng/.envs/audbcards2/lib/python3.9/site-packages/requests/models.py", line 1021, in raise_for_status                                                                                                                                                                                                                                                                                                                                                                               
    raise HTTPError(http_error_msg, response=self)                                                                                                                                                                                                                                                                                                                                                                                                                                                       
requests.exceptions.HTTPError: 401 Client Error:  for url: https://audeering.jfrog.io/artifactory/api/repositories/data-private                                                                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
The above exception was the direct cause of the following exception:   
```



This is still not conclusive, as I can leave the `data-private` tuple in and it still works - but the `all-repos` breaks it.  I have no explanation for this patterning.

Reply:
> > I added a CI job that builds the documentation and it fails when parsing the empty repo with private access rights: https://github.com/audeering/audbcards/actions/runs/7169842181/job/19521158928
> > Maybe you experience the same error?
> 
> The exception messages differ a bit:
> 
> I am getting locally:
> 
> ```
> Handler <function builder_inited at 0x7f9338f9a430> for event 'builder-inited' threw an exception (exception: An exception was raised by the backend, please see stack trace for further information.)
> ```
> 
> This is what the CI job says:
> 
> ```
> Handler <function builder_inited at 0x7f80d1412050> for event 'builder-inited' threw an exception (exception: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')))
> ```

I was able to reproduce that the process is failing also with different error messages, even though the code (#47) is identical (see [job1955205681](https://github.com/audeering/audbcards/actions/runs/7180237125/job/19552056815) and [job19552342322](https://github.com/audeering/audbcards/actions/runs/7180237125/job/19552342322)), which returned:

```
Extension error (audbcards.sphinx):
Handler <function builder_inited at 0x7f162e212050> for event 'builder-inited' threw an exception (exception: An exception was raised by the backend, please see stack trace for further information.)
Parse emodb-1.4.1... Dataset()
-- mkdir
-- header
-- dependencies
Error: Process completed with exit code 2.
```

and

```
Extension error (audbcards.sphinx):
Handler <function builder_inited at 0x7f46cc11a050> for event 'builder-inited' threw an exception (exception: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')))
Get list of available datasets... 
Error: Process completed with exit code 2.
```

In the first case `audb.available()` run through, and the process failed when calling `audb.dependencies()`. In the second case `audb.available()` failed already. So it seems that Artifactory behaves not always the same when something is not accessible. I will try to figure out at which place to best catch this.

Reply:
I managed to understand a few points and fix the documentation building issue:

### Failing with different errors and only for the mixed repos

The reason for the error we are seeing is that the public Artifactory server seems to block a user or IP for a certain amount of time after a failed or not allowed request. This means in the case of the mixed repos that the request for the valid public repo gets denied and fails. It seems also that the time for blocking the requests might differ, which explains why certain jobs failed later than others.

### Fix for the failing job

I created another repo on Artifactory called `data-empty`, which has access rights for `anonymous` user, but is simply empty. This results in the desired output and does not fail.

### Local reproducibility of the error 

I was not able to reproduce the error locally as I was providing a different user name and API key in my `~/.artifactory_python.cfg` file. When running the following command I was able to reproduce the error:

```bash
$ ARTIFACTORY_USERNAME=anonymous ARTIFACTORY_API_KEY="" python -m sphinx docs/ build/html/
```

I will create another pull request, in which I document this.

Reply:
Interesting, for whatever reason it might also [fail](https://github.com/audeering/audbcards/actions/runs/7193684686/job/19592652876?pr=45) when combining an empty repo with a public repo.
So, maybe it's just that the number of requests is too large?

Reply:
In order to avoid the error, I have now removed the `data-empty` and `all-repos` examples and show only the `data-public` example. I think for a user it should be sufficient. The other two examples where more a test to see that the functionality of the extension is working.

Reply:
I ran this approx thirty to fourty times like this:

```
NITER=10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
for i in $(seq 1 $NITER); do
    echo $i;
    echo "---------------------------------------------------";                                              
    python -m sphinx docs/ build/sphinx/html -b html                                                         
done                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
```

For me this never threw the error again.

Reply:
Good question, but I think we should solve it as part of `audbackend` and `audb`.
Hence, I would propose to stay here with the solution of using only `data-public` in the docs.